File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/Conversation/View/MSUIConversationCell.m
//
//  MSUIConversationCell.m
//  BlackFireIM
//
//  Created by benny wang on 2021/5/28.
//

#import "MSUIConversationCell.h"
#import "MSIMSDK-UIKit.h"
#import <SDWebImage.h>


@implementation MSUIConversationCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    if (self = [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
        self.backgroundColor = [UIColor d_colorWithColorLight:TCell_Nomal dark:TCell_Nomal_Dark];
        
        _headImageView = [[UIImageView alloc]init];
        _headImageView.contentMode = UIViewContentModeScaleAspectFill;
        [self.contentView addSubview:_headImageView];
        
        _timeLabel = [[UILabel alloc]init];
        _timeLabel.font = [UIFont systemFontOfSize:12];
        _timeLabel.textColor = [UIColor d_systemGrayColor];
        [self.contentView addSubview:_timeLabel];
        
        _titleLabel = [[UILabel alloc]init];
        _titleLabel.font = [UIFont boldSystemFontOfSize:17];
        _titleLabel.textColor = [UIColor d_colorWithColorLight:TText_Color dark:TText_Color_Dark];
        [self.contentView addSubview:_titleLabel];
        
        _unReadView = [[MSUnreadView alloc]init];
        [self.contentView addSubview:_unReadView];
        
        _subTitleLabel = [[UILabel alloc]init];
        _subTitleLabel.font = [UIFont systemFontOfSize:13];
        _subTitleLabel.textColor = [UIColor d_systemGrayColor];
        [self.contentView addSubview:_subTitleLabel];
        
        _genderIcon = [[UIImageView alloc]init];
        [self.contentView addSubview:_genderIcon];
        
        [self setSeparatorInset:UIEdgeInsetsMake(0, 97, 0, 0)];
        [self setSelectionStyle:UITableViewCellSelectionStyleNone];
        
        self.headImageView.layer.cornerRadius = 34;
        self.headImageView.layer.masksToBounds = YES;
    }
    return self;
}

- (void)configWithData:(MSUIConversationCellData *)convData
{
    _convData = convData;
    self.titleLabel.text = convData.title;
    self.timeLabel.text = [convData.time ms_messageString];
    self.subTitleLabel.attributedText = convData.subTitle;
    [self.unReadView setNum:convData.conv.unread_count];
    self.genderIcon.image = convData.conv.userInfo.gender == 1 ? [UIImage bf_imageNamed:@"male"] : [UIImage bf_imageNamed:@"female"];
    
    [self.headImageView sd_setImageWithURL:[NSURL URLWithString:XMNoNilString(convData.conv.userInfo.avatar)] placeholderImage:convData.avatarImage];
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    self.headImageView.frame = CGRectMake(15, 17, 68, 68);
    CGSize titleSize = [self.titleLabel sizeThatFits:CGSizeZero];
    self.titleLabel.frame = CGRectMake(self.headImageView.maxX+14, 31, MIN(titleSize.width, 150), 22);
    CGSize timeSize = [self.timeLabel sizeThatFits:CGSizeMake(200, 20)];
    self.timeLabel.frame = CGRectMake(self.contentView.width-15-timeSize.width, self.titleLabel.centerY-timeSize.height*0.5, timeSize.width, timeSize.height);
    self.subTitleLabel.frame = CGRectMake(self.titleLabel.x, self.titleLabel.maxY+5, 250, 16);
    self.unReadView.maxX = self.width-15;
    self.unReadView.y = self.subTitleLabel.y;
    
    self.genderIcon.frame = CGRectMake(self.titleLabel.maxX+5, self.titleLabel.centerY-10, 20, 20);
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/Conversation/View/MSUIConversationCell.swift
//
//  MSUIConversationCell.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/8/10.
//

import UIKit
import MSIMSDK
import Kingfisher

open class MSUIConversationCell: UITableViewCell {

    public var headImageView: UIImageView!
    
    public var titleLabel: UILabel!
    
    public var subTitleLabel: UILabel!
    
    public var timeLabel: UILabel!
    
    public var genderIcon: UIImageView!
    
    public var matchIcon: UIImageView!
    
    public var verifyIcon: UIImageView!
    
    public var unReadView: MSUnreadView!
    
    public var convData: MSUIConversationCellData?
    
    public override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        
        backgroundColor = UIColor.d_color(light: MSMcros.TCell_Nomal, dark: MSMcros.TCell_Nomal_Dark)
        
        headImageView = UIImageView()
        headImageView.contentMode = .scaleAspectFill
        contentView.addSubview(headImageView)
        
        timeLabel = UILabel()
        timeLabel.font = .systemFont(ofSize: 12)
        timeLabel.textColor = .systemGray
        contentView.addSubview(timeLabel)
        
        titleLabel = UILabel()
        titleLabel.font = .boldSystemFont(ofSize: 17)
        titleLabel.textColor = UIColor.d_color(light: MSMcros.TText_Color, dark: MSMcros.TText_Color_Dark)
        contentView.addSubview(titleLabel)
        
        unReadView = MSUnreadView()
        contentView.addSubview(unReadView)
        
        subTitleLabel = UILabel()
        subTitleLabel.font = .systemFont(ofSize: 13)
        subTitleLabel.textColor = .systemGray
        contentView.addSubview(subTitleLabel)
        
        genderIcon = UIImageView()
        contentView.addSubview(genderIcon)

        separatorInset = UIEdgeInsets(top: 0, left: 97, bottom: 0, right: 0)
        selectionStyle = .none
        
        headImageView.layer.cornerRadius = 34
        headImageView.layer.masksToBounds = true
    }
    
    public func configWithData(convData: MSUIConversationCellData) {
        
        self.convData = convData
        titleLabel.text = convData.title
        timeLabel.text = convData.time.ms_messageString()
        subTitleLabel.attributedText = convData.subTitle
        unReadView.setNum(num: convData.conv.unread_count)
            genderIcon.image = convData.conv.userInfo.gender == 1 ? UIImage.bf_imageNamed(name: "male") : UIImage.bf_imageNamed(name: "female")
        headImageView.kf.setImage(with: URL(string: convData.conv.userInfo.avatar),placeholder: convData.avatarImage)
    }
    
    open override func layoutSubviews() {
        super.layoutSubviews()
        headImageView.frame = CGRect(x: 15, y: 17, width: 68, height: 68)
        let titleSize = titleLabel.sizeThatFits(.zero)
        titleLabel.frame = CGRect(x: headImageView.right + 14, y: 31, width: min(titleSize.width, 150), height: 22)
        let timeSize = timeLabel.sizeThatFits(CGSize(width: 200, height: 20))
        timeLabel.frame = CGRect(x: contentView.width - 15 - timeSize.width, y: titleLabel.centerY - timeSize.height * 0.5, width: timeSize.width, height: timeSize.height)
        subTitleLabel.frame = CGRect(x: titleLabel.left, y: titleLabel.bottom + 5, width: 250, height: 16)
        unReadView.right = width - 15
        unReadView.top = subTitleLabel.top
            self.genderIcon.frame = CGRect(x: titleLabel.right + 5, y: titleLabel.centerY - 6, width: 12, height: 12)
    }
    
    required public init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
}
